packages/medusa/src/services/order.js

```diff
@@ -0,0 +1,447 @@
+    let updatedMethods = []
+    // partition order items to their dedicated shipping method
+    await Promise.all(
+      shipping_methods.map(async method => {
+        const { profile_id } = method
+        const profile = await this.shippingProfileService_.retrieve(profile_id)
+        // for each method find the items in the order, that are associated
+        // with the profile on the current shipping method
+        method.items = items.filter(({ content }) => {
+          if (Array.isArray(content)) {
+            // we require bundles to have same shipping method, therefore:
+            return profile.products.includes(content[0].product._id)
+          } else {
+            return profile.products.includes(content.product._id)
+          }
+        })
+        updatedMethods.push(method)
+      })
+    )
+    return updatedMethods
+  }
+
+  /**
+   * Gets an order by id.
+   * @param {string} orderId - id of order to retrieve
+   * @return {Promise<Order>} the order document
+   */
+  async retrieve(orderId) {
+    const validatedId = this.validateId_(orderId)
+    const order = await this.orderModel_
+      .findOne({ _id: validatedId })
+      .catch(err => {
+        throw new MedusaError(MedusaError.Types.DB_ERROR, err.message)
+      })
+
+    if (!order) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_FOUND,
+        `Order with ${orderId} was not found`
+      )
+    }
+    return order
+  }
+
+  /**
+   * Creates an order
+   * @param {object} order - the order to create
+   * @return {Promise} resolves to the creation result.
+   */
+  async create(order) {
+    return this.orderModel_.create(order).catch(err => {
+      throw new MedusaError(MedusaError.Types.DB_ERROR, err.message)
+    })
+  }
+
+  /**
+   * Updates an order. Metadata updates should
+   * use dedicated method, e.g. `setMetadata` etc. The function
+   * will throw errors if metadata updates are attempted.
+   * @param {string} orderId - the id of the order. Must be a string that
+   *   can be casted to an ObjectId
+   * @param {object} update - an object with the update values.
+   * @return {Promise} resolves to the update result.
+   */
+  async update(orderId, update) {
+    const validatedId = this.validateId_(orderId)
+
+    const order = await this.retrieve(orderId)
+
+    if (
+      (update.shipping_address ||
+        update.billing_address ||
+        update.payment_method ||
+        update.items) &&
+      (order.fulfillment_status !== "not_fulfilled" ||
+        order.payment_status !== "awaiting" ||
+        order.status !== "pending")
+    ) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_ALLOWED,
+        "Can't update shipping, billing, items and payment method when order is processed"
+      )
+    }
+
+    if (update.metadata) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        "Use setMetadata to update metadata fields"
+      )
+    }
+
+    const updateFields = { ...update }
```

Comment:
This would allow updates to `fulfillment_status`, `payment_status` and `status`. I think we should instead let those be read-only, so that, for example, when `createFulfillment` is called the function will automatically update `fulfillment_status`, which you already do :D 

So just make sure that we check that `update` doesn't contain any of the `...status` fields.

Reply:
You are right! We should not allow updates to `fulfillment_status` and `paymnet_status`. But the possibility of updating `status` should be there.

Reply:
I actually think we should let `status` be read-only so that it changes automatically depending of the state of the order. E.g. it would make sense to allow an `update(id, { status: "pending" })` if the order has already been fulfilled and captured? I see how it would make sense if we allowed custom order statuses, but I think it would be up to the store operators/developers to create their own statuses in `metadata`